### PR TITLE
sidh: updates multiplication and reduction mod p434 

### PR DESCRIPTION
**Pull request by armfazh**
Updates multiplication method described in:
https://github.com/microsoft/PQCrypto-SIDH/blob/v3.4/src/P434/AMD64/fp_x64_asm.S#L258

Updates reduction method described in:
https://github.com/microsoft/PQCrypto-SIDH/blob/v3.4/src/P434/AMD64/fp_x64_asm.S#L721-L727

### CODE REVIEW

File: dh/sidh/internal/templates/fp2.gotemp [hunk 1]
```diff
@@ -79,10 +79,10 @@ func mul(dest, lhs, rhs *common.Fp2) {
 	// (b - a)*(c - d) = (b*c + a*d) - a*c - b*d
 	//
 	// so (a*d + b*c) = (b-a)*(c-d) + a*c + b*d.
-	mul{{ .FIELD}}(&ac, &lhs.A, &rhs.A)       // = a*c*R*R
-	mul{{ .FIELD}}(&bd, &lhs.B, &rhs.B)       // = b*d*R*R
 	sub{{ .FIELD}}(&bMinA, &lhs.B, &lhs.A)    // = (b-a)*R
 	sub{{ .FIELD}}(&cMinD, &rhs.A, &rhs.B)    // = (c-d)*R
+	mul{{ .FIELD}}(&ac, &lhs.A, &rhs.A)       // = a*c*R*R
+	mul{{ .FIELD}}(&bd, &lhs.B, &rhs.B)       // = b*d*R*R
```
Comment (lukevalenta) [inline dh/sidh/internal/templates/fp2.gotemp:85]: What's the reason for this reordering? As far as I can tell, this is to better line up with the order of operations in the comment above: `// so (a*d + b*c) = (b-a)*(c-d) + a*c + b*d.`

Comment (armfazh) [reply dh/sidh/internal/templates/fp2.gotemp:85]: The idea is to process multiplications together. Before there was an _interruption_ for processing subtractions.

**Review by lukevalenta — APPROVED** [review]
Commit 8d9bbdc makes sense to me now, but I'm not sure how to effectively review the other two commits. Approving with that caveat, so please only merge if you're sufficiently confident in the changes or we have a second approver.

File: dh/sidh/internal/p434/arith_amd64.s [hunk 2]
```diff
@@ -22,52 +22,42 @@
 
 // Performs schoolbook multiplication of 128-bit with 256-bit
 // number. Uses MULX, ADOX, ADCX instruction.
-#define MULX128x256(IDX, M0, M1, T0, T1, T2, T3, T4, T5, T6)    \
-    XORQ    AX, AX              \
-    MOVQ    (IDX)(M0), DX       \
-    MULXQ   M1+ 0(SB), T0, T1   \ // T0 <- C0
+#define MULX128x256(I0, I1, M1, T0, T1, T2, T3, T4, T5)    \
```
Comment (wbl) [inline dh/sidh/internal/p434/arith_amd64.s:25]: It would be helpful to indicate which of these parameters are inputs and outputs and what they mean.

**Review by wbl — COMMENTED** [review]
Some more comments, or maybe a companion explaining what exactly is happening in the code would help me review. This stuff is tricky, but I think I'm starting to get the idea

**Review by claucece — APPROVED** [review]
LGTM! Thank you ;) Maybe we can add some documentation, though

File: dh/sidh/internal/p434/arith_amd64.s [hunk 3]
```diff
@@ -22,52 +22,42 @@
...
 // Performs schoolbook multiplication of 64-bit with 256-bit
 // number. Uses MULX and ADOX instructions.
 //
 // Uses registers: DX,AX
-#define MULX64x256(IDX, M0, M1, T0, T1, T2, T3, T4, T5) \
-    XORQ    AX, AX              \
-    MOVQ    (IDX)(M0), DX       \
-    MULXQ   M1+ 0(SB), T0, T1   \ // T0 <- C0
+#define MULX64x256(I0, M1, T0, T1, T2, T3, T4, T5) \
     MULXQ   M1+ 8(SB), T4, T2   \
```
[the remark this answers is not in the document]
Comment (armfazh) [reply dh/sidh/internal/p434/arith_amd64.s:66]: This documentation clarifies the calculation:

Calculates:
   (I0) x [M1][0,1,2,3] = (T0,T1,T2,T3,T4)
   |64| x |--- 256 ---| = |----- 320 ----|
Assuming the first digit multiplication was already performed.

File: dh/sidh/internal/p434/arith_amd64.s [hunk 4]
```diff
@@ -22,52 +22,42 @@
...
 
 // Performs schoolbook multiplication of 64-bit with 256-bit
 // number. Uses MULX and ADOX instructions.
 //
 // Uses registers: DX,AX
-#define MULX64x256(IDX, M0, M1, T0, T1, T2, T3, T4, T5) \
-    XORQ    AX, AX              \
-    MOVQ    (IDX)(M0), DX       \
-    MULXQ   M1+ 0(SB), T0, T1   \ // T0 <- C0
+#define MULX64x256(I0, M1, T0, T1, T2, T3, T4, T5) \
```
[the remark this answers is not in the document]
Comment (armfazh) [reply dh/sidh/internal/p434/arith_amd64.s]: I0 not used, so removed.

File: dh/sidh/internal/p434/arith_amd64.s [hunk 5]
```diff
@@ -22,52 +22,42 @@
...
     ADOXQ   AX, T4              \ // T4: interm4
     \
     XORQ    AX, AX              \
-    MOVQ    (IDX+8)(M0), DX     \
-    MULXQ   M1+ 0(SB), T5, T6   \
-    ADCXQ   T5, T1              \ // T1 <- C1
-    ADCXQ   T6, T2              \
-    \
-    MULXQ   M1+ 8(SB), T6, T5   \
+    MOVQ    I1, DX              \
```
[the remark this answers is not in the document]
Comment (armfazh) [reply dh/sidh/internal/p434/arith_amd64.s:42]: I0 not used, so removed.

File: dh/sidh/internal/p434/arith_amd64.s (hunk 2 above)
Comment (armfazh) [reply dh/sidh/internal/p434/arith_amd64.s]: removed unused params

**Pull request merged**